Update generated formulas to follow the new naming scheme

sahas-subramanian-frequenz opened the pull request:
**New power streaming methods** - _public api_


- `BatteryPool.{production_power, consumption_power}`
- `EVChargerPool.{production_power, consumption_power}`
- LogicalMeter:
  - `consumer_power`
  - `grid_production_power`
  - `grid_consumption_power`
  - `chp_power`
  - `chp_production_power`
  - `chp_consumption_power` (probably not necessary, but having just a production method was a bit weird)

**New FormulaEngine steps** - _internal_
- `ConstantValue` - for multiplying formula outputs with `-1` to get production power.
- `Clipper` - for clipping formula outputs to 0.0 in production/consumption power formulas.

**New component graph methods** - _internal_
- `is_pv_inverter`, `is_pv_meter`, `is_pv_chain`
- `is_battery_inverter`, `is_battery_meter`, `is_battery_chain`
- `is_ev_charger`, `is_ev_charger_meter`, `is_ev_charger_chain`,
- `is_chp`, `is_chp_meter`, `is_chp_chain`

These methods were particularly useful in the creation of the `consumer_power` formula.

The `*chain` methods return true if either of the other corresponding methods return true.  Maybe there's a better name, but it is for internal use.

shsms committed:
Add formula step for applying constant values

This is useful for switching the sign of values, by multiplying by -1,
for example.

Signed-off-by: Sahas Subramanian <sahas.subramanian@frequenz.com>

shsms committed:
Add a `Clipper` formula step for clipping formula output to a range

Signed-off-by: Sahas Subramanian <sahas.subramanian@frequenz.com>

shsms committed:
Add a `FormulaType` parameter for `FormulaGeneratorConfig`

Signed-off-by: Sahas Subramanian <sahas.subramanian@frequenz.com>

shsms committed:
Add `grid_{production,consumption}_power` formulas to `LogicalMeter`

Signed-off-by: Sahas Subramanian <sahas.subramanian@frequenz.com>

shsms committed:
Replace type:ignores with type asserts in formula generation methods

Signed-off-by: Sahas Subramanian <sahas.subramanian@frequenz.com>

shsms committed:
Add `LogicalMeter.chp_{,consumption,production}power` methods

Signed-off-by: Sahas Subramanian <sahas.subramanian@frequenz.com>

shsms committed:
Add `BatteryPool.{production,consumption}_power` methods

Signed-off-by: Sahas Subramanian <sahas.subramanian@frequenz.com>

shsms committed:
Add `EVChargerPool.{production,consumption}_power` methods

Signed-off-by: Sahas Subramanian <sahas.subramanian@frequenz.com>

shsms committed:
Add component graph methods to check component types

Signed-off-by: Sahas Subramanian <sahas.subramanian@frequenz.com>

shsms committed:
Add formula generator for consumer power

Signed-off-by: Sahas Subramanian <sahas.subramanian@frequenz.com>

sahas-subramanian-frequenz commented:
Need to add tests for the `consumer_power` formula,  will be ready tomorrow.

shsms committed:
Remove unused logger variable from `_logical_meter.py`

Signed-off-by: Sahas Subramanian <sahas.subramanian@frequenz.com>

cwasicki commented on src/frequenz/sdk/timeseries/_formula_engine/_formula_steps.py:
```diff
@@ -225,6 +225,41 @@ def apply(self, eval_stack: List[float]) -> None:
         eval_stack.append(self._value)
 
 
+class Clamper(FormulaStep):
```
Isn't this doing clipping instead of clampering? According to https://en.wikipedia.org/wiki/Clamper_(electronics) clampering would shift the whole series, while here the values are clipped inside a range.

cwasicki replied:
As discussed, in some programming languages the term clamping is used in the way you use it, but in electronics it is used differently. So I would favor `clip` as it's less ambiguous.

sahas-subramanian-frequenz replied:
I've renamed this,  it is internal to the sdk anyway,  so it would be fine to discuss it further at a later time  if necessary

cwasicki commented on src/frequenz/sdk/timeseries/logical_meter/_logical_meter.py:
```diff
@@ -239,3 +240,70 @@ def pv_power(self) -> FormulaEngine:
         )
         assert isinstance(engine, FormulaEngine)
         return engine
+
```
Out of curiosity, why did you choose this kind of design that you need a new function for each combination of component, metric and direction? Couldn't you save a lot of code by providing something like `power(component, direction=None)` where component is `chp`, `pv`, `grid`, etc and direction is either `consumption`, `production` or both?

sahas-subramanian-frequenz replied:
We have something like that internally,  and the public interface is just one more layer that provides easy to use functions.

Also,  the functions have to be moved to their corresponding *Pools once those are ready,  so a dynamic function will require validation to check if a specific request makes sense.  Instead we just provide functions for the possible combinations,  which won't be too many for each pool.

cwasicki commented on src/frequenz/sdk/timeseries/_formula_engine/_formula_generators/_chp_power_formula.py:
```diff
@@ -0,0 +1,107 @@
+        if not chps:
+            return None
+
+        chp_meters: set[int] = set()
```
Can the same CHP id occur multiple time in the following loop?

sahas-subramanian-frequenz replied:
nope. The component graph returns a set,  but I guess you're right, there's some unnecessary complexity here,  I will remove the intermediate list,  and directly iterate over the generator instead.

sahas-subramanian-frequenz replied:
done

cwasicki commented on src/frequenz/sdk/timeseries/ev_charger_pool/_ev_charger_pool.py:
```diff
@@ -141,9 +142,62 @@ def power(self) -> FormulaEngine:
+        method.
+
+        Returns:
+            A FormulaEngine that will calculate and stream the total power of all EV
```
`production power`

sahas-subramanian-frequenz replied:
fixed.

cwasicki commented on src/frequenz/sdk/timeseries/ev_charger_pool/_ev_charger_pool.py:
```diff
@@ -141,9 +142,62 @@ def power(self) -> FormulaEngine:
+        method.
+
+        Returns:
+            A FormulaEngine that will calculate and stream the total power of all EV
```
`consumption power`

sahas-subramanian-frequenz replied:
fixed.

cwasicki commented on src/frequenz/sdk/microgrid/_graph.py:
```diff
@@ -352,6 +495,179 @@ def validate(self) -> None:
+        successors = self.successors(component.component_id)
+        return (
+            component.category == ComponentCategory.METER
+            and len(successors) > 0
```
Can this be `==1`?

sahas-subramanian-frequenz replied:
yes,  but the component graph can be generic, doesn't have to care about the tighter requirements of the formula generator (which might go away, especially if we encounter some complicated setups)

cwasicki commented on src/frequenz/sdk/timeseries/logical_meter/_logical_meter.py:
```diff
@@ -239,3 +240,70 @@ def pv_power(self) -> FormulaEngine:
+        return engine
+
+    @property
+    def chp_production_power(self) -> FormulaEngine:
```
Looks like the methods for PV are missing in this PR?

sahas-subramanian-frequenz replied:
added.

cwasicki reviewed (commented):
Some high-level comments. Most importantly I think the methods for solar are missing.

shsms committed:
Add `LogicalMeter.pv_{production,consumption}_power` methods

Signed-off-by: Sahas Subramanian <sahas.subramanian@frequenz.com>

shsms committed:
Remove redundant test from logical meter

The battery power formula has been moved to the battery pool, and is
being tested there.

And a new test for the new PV Power methods was added in the previous
commit.

Signed-off-by: Sahas Subramanian <sahas.subramanian@frequenz.com>

cwasicki commented on src/frequenz/sdk/timeseries/logical_meter/_logical_meter.py:
```diff
@@ -170,13 +215,60 @@ def grid_current(self) -> FormulaEngine3Phase:
+        method.
+
+        Returns:
+            A FormulaEngine that will calculate and stream PV power production.
```
`PV total power`

sahas-subramanian-frequenz replied:
updated.

sahas-subramanian-frequenz commented:
Rebased on latest.

matthias-wende-frequenz commented on src/frequenz/sdk/timeseries/_formula_engine/_formula_generators/_chp_power_formula.py:
```diff
@@ -0,0 +1,105 @@
+                    f"{meter.category}. Expected ComponentCategory.METER."
+                )
+            meter_successors = component_graph.successors(meter.component_id)
+            if len(meter_successors) != 1:
```
Is that check necessary? Seems to me that we would disallow a valid grid topology, where one meter is measuring more then 1 CHPs?

shsms replied:
true,  updated.

matthias-wende-frequenz commented on src/frequenz/sdk/microgrid/_graph.py:
```diff
@@ -113,6 +113,149 @@ def successors(self, component_id: int) -> Set[Component]:
+
+    @abstractmethod
+    def is_battery_chain(self, component: Component) -> bool:
+        """Check if the specified component is part of a battery chain.
```
I suggest to use the docs to define what a battery chain is.

shsms replied:
done

matthias-wende-frequenz commented on src/frequenz/sdk/microgrid/_graph.py:
```diff
@@ -113,6 +113,149 @@ def successors(self, component_id: int) -> Set[Component]:
+
+    @abstractmethod
+    def is_ev_charger_chain(self, component: Component) -> bool:
+        """Check if the specified component is part of an EV charger chain.
```
ditto

matthias-wende-frequenz commented on src/frequenz/sdk/microgrid/_graph.py:
```diff
@@ -113,6 +113,149 @@ def successors(self, component_id: int) -> Set[Component]:
+
+    @abstractmethod
+    def is_chp_chain(self, component: Component) -> bool:
+        """Check if the specified component is part of a CHP chain.
```
ditto

matthias-wende-frequenz commented on src/frequenz/sdk/timeseries/_formula_engine/_formula_generators/_consumer_power_formula.py:
```diff
@@ -0,0 +1,121 @@
+
+
+class ConsumerPowerFormula(FormulaGenerator):
+    """Formula generator from component graph for calculating the Consumer Power."""
```
I suggest defining here what is meant by Consumer power.

shsms replied:
done

matthias-wende-frequenz commented on src/frequenz/sdk/timeseries/_formula_engine/_formula_generators/_consumer_power_formula.py:
```diff
@@ -0,0 +1,121 @@
+            return self._gen_with_grid_meter(builder, grid_meter)
+        return self._gen_without_grid_meter(builder, grid_successors)
+
+    def _gen_with_grid_meter(
```
We need to be careful here since this stops working once we add sub panels (Unterverteilungen) into the microgrid api. Maybe it's worth adding a comment here.

shsms replied:
Are we going to introduce sub panels into the component graph?  I haven't heard of any such plans.   Those are passive components right?  i.e we can't read from them or control them.

matthias-wende-frequenz replied:
As far as I remember that was the plan. @tiyash-basu-frequenz can you confirm?

matthias-wende-frequenz replied:
I suggest to add a short warning comment to the documentation such that we are aware in the future.

shsms replied:
I have added a comment: 

If the component graph supports additional types of grid successors in the future, additional checks need to be added here.

tiyash-basu-frequenz replied:
I think the term "sub panels" is misleading. If I have understood it correctly, then it is a section of the microgrid behind a dedicated fuse. I'd call it sub-section.

But I'd suggest ignoring it for now. It is not scheduled to be added to the microgrid API soon. To me even a comment seems unnecessary: because it does not improve the readability and just hints at a future functionality without knowing when we will add it. But up to you.

tiyash-basu-frequenz replied:
If you really want to track this feature, then creating an issue is the proper way to do it.

matthias-wende-frequenz replied:
You are right w.r.t. the issue. We should create one but I'd also leave a comment just to have the behavior documented because it tells that the algorithm isn't going into the depth.

tiyash-basu-frequenz replied:
Then I'd suggest to rather create the issue and add a link to the issue in the comment.

I'd suggest doing it in moderation, if never at all, because you do not want to end up adding every proposed feature in the comments.

matthias-wende-frequenz commented on src/frequenz/sdk/timeseries/_formula_engine/_formula_generators/_consumer_power_formula.py:
```diff
@@ -0,0 +1,121 @@
+            grid_meter = next(iter(grid_successors))
+            assert (
+                grid_meter.category == ComponentCategory.METER
+            ), "Only grid successor is not a meter."
```
Wouldn't it be better to raise an exception here?

shsms replied:
true, fixed.

matthias-wende-frequenz commented on src/frequenz/sdk/timeseries/_formula_engine/_formula_generators/_consumer_power_formula.py:
```diff
@@ -0,0 +1,121 @@
+        builder: ResampledFormulaBuilder,
+        grid_meter: Component,
+    ) -> FormulaEngine:
+        """Generate formula for calculating consumer power with grid-side meter.
```
We should define what a grid side meter is.

shsms replied:
maybe just calling it the grid meter makes it more obvious?

shsms replied:
renamed to "grid meter" in the docstrings.

matthias-wende-frequenz commented on src/frequenz/sdk/timeseries/_formula_engine/_formula_generators/_consumer_power_formula.py:
```diff
@@ -0,0 +1,121 @@
+        Returns:
+            A formula engine that will calculate the consumer power.
+        """
+        component_graph = connection_manager.get().component_graph
```
Why aren't you passing the compenent_graph from the calling function.

shsms replied:
because then I'd have to import `ComponentGraph` as the argument type,  which I didn't want to do.  This is just a reference,  and is required only once when creating formulas,  so there's no performance cost.

I think it is clearer to a reference from the source every time, rather than passing it around.

matthias-wende-frequenz commented on src/frequenz/sdk/timeseries/_formula_engine/_formula_generators/_consumer_power_formula.py:
```diff
@@ -0,0 +1,121 @@
+        Returns:
+            A formula engine that will calculate the consumer power.
+        """
+        component_graph = connection_manager.get().component_graph
```
ditto

matthias-wende-frequenz reviewed (commented):
Almost finished. ~I've a couple of open question w.r.t. the consumer power.~ No questions. I missed the continue in the loop over the successors :).

shsms committed:
Add formula power/current sign convention details to docstrings

Signed-off-by: Sahas Subramanian <sahas.subramanian@frequenz.com>

matthias-wende-frequenz reviewed (dismissed):
Approved apart from my last comment.
